In [1]:
import sys
import os

project_root = os.path.abspath("..")
src_path = os.path.join(project_root, "src")

if src_path not in sys.path:
    sys.path.append(src_path)


import pandas as pd
import unidecode
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score, train_test_split
from sklearn.metrics import r2_score, mean_absolute_error, mean_absolute_percentage_error

from data_analysis_octopus import (DataViz, detect_outliers_iqr, transform_outliers,
    get_varclushi, get_pca, get_kbest, get_information_value)


def count_percentage(df, columna):
    # Realizar el conteo de valores en la columna especificada
    conteo = df[columna].value_counts().reset_index()
    conteo.columns = [columna, 'conteo']
    
    # Calcular los porcentajes
    total = conteo['conteo'].sum()
    conteo['porcentaje'] = round((conteo['conteo'] / total) * 100, 2)
    
    return conteo


def create_feature_dataframe(data, column):
    feature = data.columns[0]
    category = data.at[0, column]
    conteo = data.at[0, 'conteo']
    porcentaje = data.at[0, 'porcentaje']
    
    # Constructing the final dictionary
    variable = {
        'feature': feature, 
        'category': category,
        'conteo': conteo,
        'porcentaje': porcentaje
    }
    
    # Converting the dictionary to a DataFrame
    return pd.DataFrame([variable])


def freq_discrete(df, features):
    for feature in features:
        print(f"Feature: {feature}")
        abs_ = df[feature].value_counts(dropna=False).to_frame().rename(columns={"count": "Absolute frequency"})
        rel_ = df[feature].value_counts(dropna=False, normalize= True).to_frame().rename(columns={"proportion": "Relative frequency"})
        freq = abs_.join(rel_)
        freq["Accumulated frequency"] = freq["Absolute frequency"].cumsum()
        freq["Accumulated %"] = freq["Relative frequency"].cumsum()
        freq["Absolute frequency"] = freq["Absolute frequency"].map(lambda x: "{:,.0f}".format(x))
        freq["Relative frequency"] = freq["Relative frequency"].map(lambda x: "{:,.2%}".format(x))
        freq["Accumulated frequency"] = freq["Accumulated frequency"].map(lambda x: "{:,.0f}".format(x))
        freq["Accumulated %"] = freq["Accumulated %"].map(lambda x: "{:,.2%}".format(x))
        display(freq)


def clean_name_columns(columns):
    cleaned_columns_dict = {}
    for col in columns:
        
        processed_col = unidecode.unidecode(
            col.lower()
               .replace("¿", "")
               .replace("?", "")
               .replace(" (numérica)", "")
               .replace(" (numérico)", "")
               .replace(" ", "_")
        )

        cleaned_columns_dict[col] = processed_col

    return cleaned_columns_dict


def classify_colors(color):

    if color != "azul":
        return "no es azul"
    
    return color


def classify_pets(pet):

    if pet in ["perrro", "perro", "perros"]:
        return "perro"
    
    if pet in ["gato"]:
        return "gato"
    
    return "otro"


def classify_tattoos(n_tattoos):
    if n_tattoos <= 0:
        return "sin tatuajes"
    
    if n_tattoos > 0:
        return "con tatuajes"
    
    return float('nan')


def classify_siblings(siblings):

    if color != "azul":
        return "no es azul"
    
    return color



def clean_text(text):
    text = unidecode.unidecode(text.lower())
    return text

# Regresión lineal

## Lectura de los datos

In [2]:
filename = "/workspace/data/Team-calor- o-team-frío.csv"
raw_df = pd.read_csv(filename).drop(columns="Marca temporal")

new_name_columns = clean_name_columns(raw_df.columns)
raw_df = raw_df.rename(columns=new_name_columns)
raw_df.head()

color_primario_favorito edad_en_anos estatura_en_metros  \
0                    Azul           29               1.92   
1                    Azul           26               1.57   
2                    Azul           25               1.71   
3                    Rojo           27               1.78   
4                    Azul           23                1.7   

  tipo_de_personalidad numero_de_vasos_de_agua_que_tomas_al_dia dia_o_noche  \
0         Introvertido                                        6       Noche   
1         Introvertido                                        5       Noche   
2         Introvertido                                        3       Noche   
3         Introvertido                                        4       Noche   
4         Introvertido                                        4       Noche   

  actividad_fisica mascota_favorita chile_del_que_pica_o_del_que_no_pica  \
0     ¿Qué es eso?           Perrro                         Del que pica   
1     ¿Qué es eso?           Perrro                         Del que pica   
2               Si           Perrro                         Del que pica   
3     ¿Qué es eso?           Perrro                         Del que pica   
4               Si             Gato                      Del que no pica   

  numero_de_hermanos las_quesadillas_van_con_queso numero_de_tatuajes  \
0                  2                     Con queso                  0   
1                  0                     Con queso                  0   
2                  1                     Con queso                  0   
3                  1                     Sin queso                  0   
4                  2                     Con queso                  0   

  team_frio_o_team_calor  
0                    NaN  
1                    NaN  
2                    NaN  
3                    NaN  
4                    NaN

## Preprocesamiento

In [3]:
categoricas = [
    "color_primario_favorito",
    "tipo_de_personalidad",
    "dia_o_noche",
    "actividad_fisica",
    "mascota_favorita",
    "chile_del_que_pica_o_del_que_no_pica",
    "las_quesadillas_van_con_queso",
    "team_frio_o_team_calor"
]

numericas = [
    "edad_en_anos",
    "numero_de_hermanos",
    "estatura_en_metros",
    "numero_de_vasos_de_agua_que_tomas_al_dia",
    "numero_de_tatuajes"
]

In [4]:
processed_df = raw_df.copy()

processed_df[categoricas] = processed_df[categoricas].fillna("")
processed_df = processed_df.applymap(lambda value: value.strip())
processed_df[categoricas] = processed_df[categoricas].applymap(clean_text)

# Reclasificaicón de categorías
processed_df["color_primario_favorito"] = processed_df["color_primario_favorito"].apply(classify_colors)
processed_df["mascota_favorita"] = processed_df["mascota_favorita"].apply(classify_pets)
# Se transforman a con tatuajes o sin tatuajes, ya que, al detectar outliers, 
# mediante IQR, aquellos con valores mayoes a 0, son valores atípicos, 
# para conservar la información se recategoriza en: 'con tatuajes' & 'sin tatuajes
processed_df["tatuajes"] = processed_df["numero_de_tatuajes"].apply(pd.to_numeric, errors='coerce').apply(classify_tattoos)
# processed_df["hermanos"] = processed_df["numero_de_tatuajes"].apply(pd.to_numeric, errors='coerce').apply(classify_tattoos)


/tmp/ipykernel_111807/2966697770.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  processed_df = processed_df.applymap(lambda value: value.strip())
/tmp/ipykernel_111807/2966697770.py:5: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  processed_df[categoricas] = processed_df[categoricas].applymap(clean_text)


In [5]:
categoricas.append("tatuajes")

In [6]:
freq_discrete(processed_df, categoricas)

Feature: color_primario_favorito


Absolute frequency Relative frequency  \
color_primario_favorito                                         
azul                                    69             57.98%   
no es azul                              50             42.02%   

                        Accumulated frequency Accumulated %  
color_primario_favorito                                      
azul                                       69        57.98%  
no es azul                                119       100.00%

Feature: tipo_de_personalidad


Absolute frequency Relative frequency  \
tipo_de_personalidad                                         
introvertido                         74             62.18%   
extrovertido                         45             37.82%   

                     Accumulated frequency Accumulated %  
tipo_de_personalidad                                      
introvertido                            74        62.18%  
extrovertido                           119       100.00%

Feature: dia_o_noche


Absolute frequency Relative frequency Accumulated frequency  \
dia_o_noche                                                               
noche                       85             71.43%                    85   
dia                         34             28.57%                   119   

            Accumulated %  
dia_o_noche                
noche              71.43%  
dia               100.00%

Feature: actividad_fisica


Absolute frequency Relative frequency Accumulated frequency  \
actividad_fisica                                                               
si                               84             70.59%                    84   
?que es eso?                     33             27.73%                   117   
no                                2              1.68%                   119   

                 Accumulated %  
actividad_fisica                
si                      70.59%  
?que es eso?            98.32%  
no                     100.00%

Feature: mascota_favorita


Absolute frequency Relative frequency Accumulated frequency  \
mascota_favorita                                                               
perro                            80             67.23%                    80   
gato                             32             26.89%                   112   
otro                              7              5.88%                   119   

                 Accumulated %  
mascota_favorita                
perro                   67.23%  
gato                    94.12%  
otro                   100.00%

Feature: chile_del_que_pica_o_del_que_no_pica


Absolute frequency Relative frequency  \
chile_del_que_pica_o_del_que_no_pica                                         
del que pica                                         87             73.11%   
del que no pica                                      32             26.89%   

                                     Accumulated frequency Accumulated %  
chile_del_que_pica_o_del_que_no_pica                                      
del que pica                                            87        73.11%  
del que no pica                                        119       100.00%

Feature: las_quesadillas_van_con_queso


Absolute frequency Relative frequency  \
las_quesadillas_van_con_queso                                         
con queso                                     90             75.63%   
sin queso                                     29             24.37%   

                              Accumulated frequency Accumulated %  
las_quesadillas_van_con_queso                                      
con queso                                        90        75.63%  
sin queso                                       119       100.00%

Feature: team_frio_o_team_calor


Absolute frequency Relative frequency  \
team_frio_o_team_calor                                         
team frio                              87             73.11%   
team calor                             18             15.13%   
                                       14             11.76%   

                       Accumulated frequency Accumulated %  
team_frio_o_team_calor                                      
team frio                                 87        73.11%  
team calor                               105        88.24%  
                                         119       100.00%

Feature: tatuajes


Absolute frequency Relative frequency Accumulated frequency  \
tatuajes                                                                   
sin tatuajes                 93             78.15%                    93   
con tatuajes                 25             21.01%                   118   
NaN                           1              0.84%                   119   

             Accumulated %  
tatuajes                    
sin tatuajes        78.15%  
con tatuajes        99.16%  
NaN                100.00%

In [7]:
processed_df[numericas] = processed_df[numericas].apply(pd.to_numeric, errors='coerce')

#### Detección y remoción de variables poco pobladas

In [8]:
# No es necesario eliminar variables poco pobladas
THRESHOLD = 65
completitud_df = DataViz.completitud(processed_df)
completitud_df[completitud_df["% valores nulos"] >= THRESHOLD]

Empty DataFrame
Columns: [Total de nulos, % valores nulos]
Index: []

#### Detección y remoción de valores extremos

In [9]:
for c in numericas:
    lower_bound, upper_bound = detect_outliers_iqr(processed_df, c)
    processed_df = transform_outliers(processed_df, c, lower_bound, upper_bound)

#### Detección y tratamiento de valores ausentes

In [10]:
for col in numericas:
    media = processed_df[col].median()
    processed_df.loc[:, col] = processed_df[col].fillna(media)  

for col in categoricas:
    moda = processed_df[col].mode()[0]
    processed_df.loc[:, col] = processed_df[col].fillna(moda)

In [11]:
processed_df[numericas].describe()

edad_en_anos  numero_de_hermanos  estatura_en_metros  \
count    119.000000          119.000000          119.000000   
mean      24.756303            1.634454            1.749118   
std        4.085871            0.908289            0.149020   
min       14.500000           -0.500000            1.400000   
25%       22.000000            1.000000            1.640000   
50%       24.000000            2.000000            1.715000   
75%       27.000000            2.000000            1.800000   
max       34.500000            3.500000            2.040000   

       numero_de_vasos_de_agua_que_tomas_al_dia  numero_de_tatuajes  
count                                119.000000               119.0  
mean                                   6.327731                 0.0  
std                                    2.690565                 0.0  
min                                   -2.000000                 0.0  
25%                                    4.000000                 0.0  
50%                                    6.000000                 0.0  
75%                                    8.000000                 0.0  
max                                   14.000000                 0.0

#### Análisis de correlación

In [12]:
processed_df[numericas].corr()

edad_en_anos  numero_de_hermanos  \
edad_en_anos                                  1.000000            0.137353   
numero_de_hermanos                            0.137353            1.000000   
estatura_en_metros                            0.108485            0.171968   
numero_de_vasos_de_agua_que_tomas_al_dia      0.001545            0.077180   
numero_de_tatuajes                                 NaN                 NaN   

                                          estatura_en_metros  \
edad_en_anos                                        0.108485   
numero_de_hermanos                                  0.171968   
estatura_en_metros                                  1.000000   
numero_de_vasos_de_agua_que_tomas_al_dia            0.157770   
numero_de_tatuajes                                       NaN   

                                          numero_de_vasos_de_agua_que_tomas_al_dia  \
edad_en_anos                                                              0.001545   
numero_de_hermanos                                                        0.077180   
estatura_en_metros                                                        0.157770   
numero_de_vasos_de_agua_que_tomas_al_dia                                  1.000000   
numero_de_tatuajes                                                             NaN   

                                          numero_de_tatuajes  
edad_en_anos                                             NaN  
numero_de_hermanos                                       NaN  
estatura_en_metros                                       NaN  
numero_de_vasos_de_agua_que_tomas_al_dia                 NaN  
numero_de_tatuajes                                       NaN

#### Detección y remoción de variables unitarias (unarias)
cuando el 90% de la información se agrupe dentro una sola variable

In [13]:
dfs_list = [ ]
for c in processed_df.columns:
    tmp_count = count_percentage(processed_df, c)
    
    tmp_df = tmp_count[tmp_count["porcentaje"] >= 90]

    if not tmp_df.empty:
        dfs_list.append(create_feature_dataframe(tmp_df, c))

unit_vars = pd.concat(dfs_list).reset_index(drop=True)
unit_vars

feature  category  conteo  porcentaje
0  numero_de_tatuajes       0.0     119       100.0

In [14]:
processed_df = processed_df.drop(columns=unit_vars["feature"].to_list())

#### Tratamiento de variables categoricas

In [15]:
dummies_df = pd.get_dummies(processed_df[categoricas], drop_first=True, dtype=int)
processed_df = pd.concat([processed_df, dummies_df], axis=1)

#### Escalamiento de los datos

In [16]:
mask_columns = processed_df.columns.isin(categoricas) == False
subset_df = processed_df.loc[:, mask_columns]

minmax = MinMaxScaler()
minmax_df = pd.DataFrame(
    minmax.fit_transform(subset_df),
    columns=subset_df.columns
)

X = minmax_df.loc[:, minmax_df.columns != "estatura_en_metros"]
y = minmax_df[["estatura_en_metros"]]


### Selección de variables

#### Clustering de variables

In [17]:
vc_df, vc_rsquare_df, best_features = get_varclushi(X)
best_features_df = best_features.to_frame().reset_index(drop=True)
best_features_df

Variable
0  las_quesadillas_van_con_queso_sin queso
1                       numero_de_hermanos
2       color_primario_favorito_no es azul
3                      actividad_fisica_si
4                      actividad_fisica_no
5                        dia_o_noche_noche

In [18]:
top_features_vc = best_features_df.head(5)["Variable"].to_list()

## Kbest

In [19]:
scores_df = get_kbest(X, y, k=10)
scores_df

Feature     Score
10  chile_del_que_pica_o_del_que_no_pica_del que pica  4.499563
13                   team_frio_o_team_calor_team frio  3.874261
2                                  numero_de_hermanos  3.565464
3                  color_primario_favorito_no es azul  3.240512
11            las_quesadillas_van_con_queso_sin queso  3.192708
12                  team_frio_o_team_calor_team calor  3.022826
1            numero_de_vasos_de_agua_que_tomas_al_dia  2.986635
7                                 actividad_fisica_si  2.726070
0                                        edad_en_anos  1.393381
5                                   dia_o_noche_noche  1.296043
4                   tipo_de_personalidad_introvertido  0.879603
8                               mascota_favorita_otro  0.446454
14                              tatuajes_sin tatuajes  0.306893
9                              mascota_favorita_perro  0.022757
6                                 actividad_fisica_no  0.000071

In [20]:
top_features_kbest = scores_df.head(5)["Feature"].to_list()

### Modelado

##### Selección de mejores features

In [21]:
best_features = list(set(top_features_vc + top_features_kbest))
X = processed_df[best_features]
y = processed_df[["estatura_en_metros"]]

X.head()

actividad_fisica_si  color_primario_favorito_no es azul  \
0                    0                                   0   
1                    0                                   0   
2                    1                                   0   
3                    0                                   1   
4                    1                                   0   

   actividad_fisica_no  chile_del_que_pica_o_del_que_no_pica_del que pica  \
0                    0                                                  1   
1                    0                                                  1   
2                    0                                                  1   
3                    0                                                  1   
4                    0                                                  0   

   team_frio_o_team_calor_team frio  las_quesadillas_van_con_queso_sin queso  \
0                                 0                                        0   
1                                 0                                        0   
2                                 0                                        0   
3                                 0                                        1   
4                                 0                                        0   

   numero_de_hermanos  
0                 2.0  
1                 0.0  
2                 1.0  
3                 1.0  
4                 2.0

#### Train-test split

In [22]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

In [23]:
scaler_X = MinMaxScaler()
scaler_y = MinMaxScaler()

scaler_X.fit(X_train)
scaler_y.fit(y_train.values.reshape(-1, 1))

# Transformar los datos de entrenamiento y prueba
scaled_X_train = scaler_X.transform(X_train)
scaled_X_test = scaler_X.transform(X_test)
scaled_y_train = scaler_y.transform(y_train.values.reshape(-1, 1))
scaled_y_test = scaler_y.transform(y_test.values.reshape(-1, 1))

In [24]:
linreg = LinearRegression()
linreg.fit(scaled_X_train, scaled_y_train)
linreg.score(scaled_X_train, scaled_y_train)

0.17467744612339986

#### Cross-validation

In [25]:
ls_res = cross_val_score(X=scaled_X_train, y=scaled_y_train, estimator=linreg, cv=4, scoring="r2", n_jobs=-1)
ls_res

array([-0.02238477,  0.05827925,  0.15724925, -0.03745985])

In [26]:
ls_res.mean(), ls_res.std()

(0.03892096960402108, 0.07740921638171061)

#### Validación en test

In [27]:
linreg.score(scaled_X_test, scaled_y_test)

0.08065834043938713

#### Predicción de test

In [28]:
scaled_y_pred = linreg.predict(scaled_X_test)

In [29]:
pred_df = pd.DataFrame(scaler_y.inverse_transform(scaled_y_pred), columns=["y_pred"])
pred_df.head()

y_pred
0  1.794437
1  1.794437
2  1.778950
3  1.759122
4  1.714734

### Métricas de performance

#### R²

In [30]:
r2_score(y_true=scaled_y_test, y_pred=scaled_y_pred)

0.08065834043938713

#### PEMA

In [31]:
pema = mean_absolute_percentage_error(y_true=scaled_y_test, y_pred=scaled_y_pred)
f"{pema:,.2%}"

'37.36%'

#### MAE

In [32]:
mean_absolute_error(y_true=scaled_y_test, y_pred=scaled_y_pred)

0.17604330720812364

#### Interpretación

In [33]:
linreg.intercept_

array([0.24947707])

In [34]:
df_res = pd.DataFrame(data=zip(X_train.columns, linreg.coef_[0]), columns=["feature", "weight"])
df_res["weight"] = df_res["weight"].round(8)

df_res.sort_values(by="weight", ascending=False)

feature    weight
6                                 numero_de_hermanos  0.184948
4                   team_frio_o_team_calor_team frio  0.124535
5            las_quesadillas_van_con_queso_sin queso  0.095359
0                                actividad_fisica_si  0.074440
3  chile_del_que_pica_o_del_que_no_pica_del que pica  0.052263
2                                actividad_fisica_no  0.000000
1                 color_primario_favorito_no es azul -0.024198

### Conclusiones

En cuanto al desempeño, tanto R² (8%), MAE (0.17) noS dicen que el modelo explican muy poca variabilidad de la estatura y hay poca precisión en las predicciones. En cuanto al error absoluto medio porcentual (MAPE) es del 37.36%, lo que indica una diferncia significativa entre las predicciones y los valores reales de la estatura. <br><br>
En cuanto a las características, el número de hermanos y team (frío o calor), son las caracetrisitcas con mayor influencia, con coeficientes más altos, en comparación con otros. Tener más hermanos y ser team frío están asociados a una estatura ligereamente mayor. <br><br>
El modelo de regresión lineal tiene un desempeño limitado para inferir la estatura. Me parece que en se debe, en parte, a que las características utilizadas no tienen un impacto significativo en la estatura, o no son causales de ser más pequeños o altos.



